CARDS-1359: Integrate SAML Authentication

IntegralProgrammer opened the pull request:
The PR integrates the `sling-org-apache-sling-auth-saml2` bundle so that CARDS users may authenticate themselves using a trusted Identity Provider (IdP) server.


To test, follow the instructions in `Utilities/Administration/SAML/SAML_SETUP.md` - specifically those for locally testing using a Keycloak Docker container. Ensure that you have also completed the _Security Tests_.

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Add Apache Sling SAML2 Handler OSGi bundle

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Add scripts for configuring SAML authentication on a running CARDS instance

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Add --saml option to start_cards.sh

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Add a script for creating the keystore

Add documentation and an example Keycloak realm JSON file

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Add documentation for security testing

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Upgrade Java version used within the Docker container from 8 to 11 as the SAML module requires Java 11

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Add the --saml option to the generate_compose_yaml.py script to instruct the Java process within the Docker container to enable the SAML OSGi bundle

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

When the --saml option for the generate_compose_yaml.py script is specified, mount (as read-only) compose-cluster/samlKeystore.p12 in the container as /opt/cards/samlKeystore.p12

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

When the --composum option for the generate_compose_yaml.py script is specified, enable the Composum browser at /bin/browser.html in the CARDS Docker container

Composum is required for SAML configuration

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

[misc] Clean up tabs+spaces formatting in generate_compose_yaml.py. Use two spaces for an indent

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Update .gitignore files

IntegralProgrammer commented:
## Updated Testing Instructions

- Follow the instructions in `Utilities/Administration/SAML/SAML_SETUP.md` and ensure that the SAML setup still works.
- Repeat the instructions in `Utilities/Administration/SAML/SAML_SETUP.md` but this time, instead of starting CARDS with the `start_cards.sh` script, start it using Docker Compose by doing the following:
  - `cd compose-cluster`
  - `python3 generate_compose_cluster.py --oak_filesystem --saml --composum`
  - `cp ../samlKeystore.p12 .`
  - `docker-compose up -d`
- This PR also changes the CARDS container's version of Java from 8 to 11 (as is required by the SAML module) and therefore doing some basic smoke testing of various functionalities might be advisable.

IntegralProgrammer committed:
CARDS-1540: When SAML authentication is enabled, the permission scheme should default to "trusted"

If start_cards.sh is called with the --saml flag, default to the "trusted" permissions mode instead of "open"

IntegralProgrammer committed:
CARDS-1540: When SAML authentication is enabled, the permission scheme should default to "trusted"

If the CARDS Docker container is started with the SAML_AUTH_ENABLED environment variable set to "true", default to the "trusted" permissions mode instead of "open"

acrowthe reviewed (approved):
Tested: Security tests work as expected, both on docker and non-docker builds. If the user signs out of cards and the saml session is ended using the keycloak admin user, you are properly asked to re-login. Haven't seen any repeats of the error page from previous testing.

sdumitriu commented on Utilities/Administration/SAML/.gitignore:
```diff
@@ -0,0 +1,4 @@
+samlKeystore.p12
+samlSPcert.pem
+samlSPkey.pem
+signingCert.pem
```
I would ignore all `*.pem` files, I don't see any usecase for allowing pems to be committed.

```suggestion
*.p12
*.key
*.pem
```

Actually I would add this in the main `.gitignore`.

IntegralProgrammer replied:
We can't ignore all `*.pem` files in the repository as the file `compose-cluster/AppleIncRootCertificate.pem` needs to be checked in. Instead, what I have done is ignore all `*.p12`, `*.key`, and `*.pem` files in the `Utilities/Administration/SAML` directory and the file `samlKeystore.p12` in the root `.gitignore`.

See commit `a8e5f36`

sdumitriu replied:
Ignored files are not completely discarded by git, but instead they are not listed (by default, adding `--ignored` to the relevant git commands will still show them). Explicitly added files, even if matching an ignore pattern, will still be tracked by git just like if they aren't ignored.

sdumitriu commented on Utilities/Administration/SAML/SAML_SETUP.md:
````diff
@@ -0,0 +1,110 @@
+If you are testing locally, you may use a Keycloak Docker container
+
+```bash
+docker pull quay.io/keycloak/keycloak:15.0.2
````
Should the version be hardcoded, or use the latest?

IntegralProgrammer replied:
I have addressed this by specifying the _latest_ version of KeyCloak to be used and then specifying that version `15.0.2` was used during the creation of these instructions.

See commit `583fc4b`.

sdumitriu commented on Utilities/Administration/SAML/setup_saml.sh:
```diff
@@ -0,0 +1,25 @@
+
+python3 create_service_user_mapping.py || { echo "Failed to create service user mapping"; exit -1; }
+python3 create_saml_service_user.py || { echo "Failed to create saml service user"; exit -1; }
+python3 configure_saml_service_user_permissions.py || { echo "Failed to configure saml service user permissions"; exit -1; }
```
These 3 scripts (`create_saml_service_user`, `create_service_user_mapping`, `configure_saml_service_user_permissions`), don't seem to have any user-configured parameters in them. I would instead embed the configuration in the feature file, like in [here](https://github.com/data-team-uhn/cards/blob/fd5898421fa1582c99f51aaf75b2cc65c4d55eb6/modules/token-authentication/src/main/features/feature.json#L25-L38).

IntegralProgrammer replied:
Added in commit `baa18d9`.

sdumitriu commented on Utilities/Administration/SAML/setup_saml.sh:
```diff
@@ -0,0 +1,25 @@
+python3 create_service_user_mapping.py || { echo "Failed to create service user mapping"; exit -1; }
+python3 create_saml_service_user.py || { echo "Failed to create saml service user"; exit -1; }
+python3 configure_saml_service_user_permissions.py || { echo "Failed to configure saml service user permissions"; exit -1; }
+python3 add_saml_sp_config.py || { echo "Failed to add SAML SP configuration"; exit -1; }
```
This script has a lot of parameters that may be needed for a proper configuration, and running it like this from within this script makes them inaccessible. If the other 3 scripts are replaced by static configuration, this file will be redundant, all that's needed is to run the `add_saml_sp_config` script with the right arguments.

IntegralProgrammer replied:
Done in commit `6ba9a90`.

sdumitriu commented on Utilities/Administration/SAML/add_saml_sp_config.py:
```diff
@@ -0,0 +1,82 @@
+form_data["saml2SessionAttr"] = "saml2AuthInfo"
+form_data["saml2IDPDestination"] = args.saml2IDPDestination
+form_data["saml2LogoutURL"] = args.saml2LogoutURL
+form_data["saml2SPEnabled"] = ["true","false"]
+form_data["saml2SPEncryptAndSign"] = ["true", "false"]
```
Um, why do these two have both `true` and `false` values?

IntegralProgrammer replied:
I'm not sure either why this happens but based on inspecting the browser's _Network Monitor_ when making configuration changes in the _Apache Sling Web Console Configuration_ (`/system/console/configMgr`) it seems that the convention is to send `["true", "false"]` when a checkbox is ticked and `"false"` when it is not.

This is what happens for the SAML configuration settings panel as well as for the _Apache Sling Create User_ settings panel and probably for many others.

sdumitriu replied:
I see, kind of makes sense.

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Address review comment - remove the configure_saml_service_user_permissions.py, create_saml_service_user.py, and create_service_user_mapping.py scripts and place these configurations in the base.json feature file instead.

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Address review comment - remove the (now) redundant setup_saml.sh script and update the SAML_SETUP.md documentation file

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Address review comment - update SAML_SETUP.md and specify either to use the latest version of the Keycloak Docker image (quay.io/keycloak/keycloak) or the specific version that was used during the creation of this documentation (quay.io/keycloak/keycloak:15.0.2)

IntegralProgrammer committed:
CARDS-1359: Integrate SAML authentication

Address review comment - revise .gitignore files for ignoring SAML key files
